### PR TITLE
[AIRFLOW-1978] WinRM windows operator and hook

### PR DESCRIPTION
**Pull request by cloneluke**
Make sure you have checked _all_ steps below.

### JIRA
- [AIRFLOW-1978] My PR addresses the following
    - https://issues.apache.org/jira/browse/AIRFLOW-1978

### Description
- [X] This PR allows airflow DAGs to connect and run remote windows commands via winRM protocol. It uses https://github.com/diyan/pywinrm / https://pypi.python.org/pypi/pywinrm

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [X] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [ ] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by cloneluke** [comment]
@r39132  and @Fokko  ^

**Comment by cloneluke** [comment]
replaces old PR https://github.com/apache/incubator-airflow/pull/2923, I've squashed commits, added to the docs and resolved a conflict

**Comment by cloneluke** [comment]
also merged with 1.10 codebase

File: airflow/contrib/example_dags/example_winrm_operator.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+# Caveat: This Dag will not run because of missing scripts.
+# The purpose of this is to give you a sample of a real world example DAG!
+# --------------------------------------------------------------------------------
+
+# --------------------------------------------------------------------------------
+# Load The Dependencies
+# --------------------------------------------------------------------------------
+import airflow
+from builtins import range
+#from airflow.operators.bash_operator import BashOperator
```
Comment (r39132) [inline airflow/contrib/example_dags/example_winrm_operator.py]: Remove this line

File: airflow/contrib/example_dags/example_winrm_operator.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+
+args = {
+    'owner': 'airflow',
+    'start_date': airflow.utils.dates.days_ago(2)
+}
+
+dag = DAG(
+    dag_id='POC_winrm_parallel', default_args=args,
+    schedule_interval='46 23 21 11 *',
```
Comment (r39132) [inline airflow/contrib/example_dags/example_winrm_operator.py]: Any reason for this type of schedule interval? Why not something more straight-forward like `0 0 * * *`?

File: docs/code.rst [hunk 3]
```diff
@@ -179,6 +179,7 @@ Operators
 .. autoclass:: airflow.contrib.operators.spark_submit_operator.SparkSubmitOperator
 .. autoclass:: airflow.contrib.operators.sqoop_operator.SqoopOperator
 .. autoclass:: airflow.contrib.operators.ssh_operator.SSHOperator
+.. autoclass:: airflow.contrib.operators.winrm_operator.WinRMOperator
```
Comment (r39132) [inline docs/code.rst]: This should be in alphabetic ordering... and V comes before W

File: docs/code.rst [hunk 4]
```diff
@@ -381,6 +382,7 @@ Community contributed hooks
 .. autoclass:: airflow.contrib.hooks.vertica_hook.VerticaHook
 .. autoclass:: airflow.contrib.hooks.spark_jdbc_hook.SparkJDBCHook
```
Comment (r39132) [inline docs/code.rst]: I know this is not your entry... but it looks like this is not in order either.

File: setup.py [hunk 5]
```diff
@@ -194,6 +195,7 @@ def write_version(filename=os.path.join(*['airflow',
     'rednose',
     'paramiko',
     'pysftp',
+    'pywinrm',
```
Comment (r39132) [inline setup.py]: Hmn.. looks like the others are no in alphabetic ordering.. can you possibly fix these?

File: setup.py [hunk 6]
```diff
@@ -304,6 +306,7 @@ def do_setup():
             'sendgrid' : sendgrid,
             'slack': slack,
             'ssh': ssh,
+            'winrm': winrm,
```
Comment (r39132) [inline setup.py]: Add your entry at the correct place to ensure alphabetic ordering.

**Review by r39132 — CHANGES_REQUESTED** [review]

**Comment by r39132** [comment]
Also... you do need to squash all of your commits into 1.

**Comment by r39132** [comment]
@cloneluke see comments above.

**Comment by r39132** [comment]
@cloneluke last reminder.. else I will close this PR. Please address the comments.

File: docs/code.rst (hunk 4 above)
Comment (r39132) [reply docs/code.rst]: Ad a comment for future developers to maintain the alphabetic ordering of these entries.

File: airflow/contrib/example_dags/example_winrm_operator.py (hunk 2 above)
Comment (cloneluke) [reply airflow/contrib/example_dags/example_winrm_operator.py]: fixed, commit coming

File: docs/code.rst (hunk 3 above)
Comment (cloneluke) [reply docs/code.rst]: fixed, commit coming

File: docs/code.rst (hunk 4 above)
Comment (cloneluke) [reply docs/code.rst]: fixed, commit coming

File: setup.py (hunk 5 above)
Comment (cloneluke) [reply setup.py]: fixed, commit coming

File: setup.py (hunk 6 above)
Comment (cloneluke) [reply setup.py]: fixed, commit coming

**Comment by cloneluke** [comment]
@r39132  please review my changes you requested and let me know

File: airflow/contrib/example_dags/example_winrm_operator.py (hunk 1 above)
Comment (cloneluke) [reply airflow/contrib/example_dags/example_winrm_operator.py]: fixed

**Comment by r39132** [comment]
Thx. I will squash these when I merge.

**Pull request closed without merging**